public/app/features/dashboard/state/runRequest.ts

```diff
@@ -0,0 +1,150 @@
+// Libraries
+import { Observable, of, timer, merge } from 'rxjs';
+import { flatten, map as lodashMap, isArray, isString } from 'lodash';
+import { map, catchError, takeUntil, mapTo, share } from 'rxjs/operators';
+
+// Utils & Services
+import { LoadingState, dateMath } from '@grafana/data';
+
+// Types
+import { DataSourceApi, DataQueryRequest, PanelData, DataQueryResponsePacket, DataQueryError } from '@grafana/ui';
+
+type MapOfResponsePackets = { [str: string]: DataQueryResponsePacket };
+
+interface RunningQueryState {
+  packets: { [key: string]: DataQueryResponsePacket };
+  panelData: PanelData;
+}
+
+/*
+ * This function should handle composing a PanelData from multiple responses
+ * Does not handle deltas yet
+ */
+export function processResponsePacket(packet: DataQueryResponsePacket, state: RunningQueryState): RunningQueryState {
+  const request = state.panelData.request;
+  const packets: MapOfResponsePackets = {
+    ...state.packets,
+  };
+
+  packets[packet.key || 'A'] = packet;
+
+  // Update the time range
+  let timeRange = request.range;
+  if (isString(timeRange.raw.from)) {
+    timeRange = {
+      from: dateMath.parse(timeRange.raw.from, false),
+      to: dateMath.parse(timeRange.raw.to, true),
+      raw: timeRange.raw,
+    };
+  }
+
+  const combinedData = flatten(
+    lodashMap(packets, (packet: DataQueryResponsePacket) => {
+      return packet.data;
+    })
+  );
+
+  const panelData = {
+    state: LoadingState.Done,
```

Comment:
how should we know Done?  response is sent before unsubscribe... right?

Reply:
yea, for simple cases Done will be set automatically on first response. So data sources does not have to think about this (for simple cases). For streaming/partial results I am not sure. Think easiest is just to add loading state in the DataQueryResponePacket to data sources can communicate back if more is coming.